libs/editor/egui_editor/src/events.rs

```diff
@@ -1,97 +1,306 @@
+use crate::buffer::Buffer;
 use crate::cursor::Cursor;
-use crate::cursor_types::DocCharOffset;
-use crate::editor::Editor;
-use egui::{Event, Key, PointerButton, Ui, Vec2};
+use crate::debug::DebugInfo;
+use crate::galleys::Galleys;
+use crate::offset_types::DocCharOffset;
+use crate::unicode_segs;
+use crate::unicode_segs::UnicodeSegs;
+use egui::{Event, Key, PointerButton, Vec2};
+use std::iter;
 use std::ops::Range;
+use unicode_segmentation::UnicodeSegmentation;
 
-impl Editor {
-    pub fn key_events(&mut self, ui: &mut Ui) {
-        if !self.galleys.is_empty() {
-            let (galley_idx, _) = self.galley_and_cursor();
-            self.fix_cursor(galley_idx, false);
-        }
+/// represents a modification made as a result of event processing
+#[derive(Debug)]
+enum Modification<'a> {
+    Cursor { cur: Cursor },   // modify the cursor state
+    Insert { text: &'a str }, // insert text at cursor location
+    Delete,                   // delete selection or character before cursor
+    DebugToggle,              // toggle debug overlay
+}
+
+/// processes `events` and returns a boolean pair representing (text_updated, selection_updated)
+pub fn process(
+    events: &[Event], galleys: &Galleys, buffer: &mut Buffer, segs: &mut UnicodeSegs,
+    cursor: &mut Cursor, debug: &mut DebugInfo,
+) -> (bool, bool) {
+    let modifications = calc_modifications(events, segs, galleys, buffer, *cursor);
+    apply_modifications(modifications, buffer, segs, cursor, debug)
```

Comment:
two-step process clarifies which parts of the code in this file reference/mutate what inputs